Add Named errors tx timed out

manoranjith opened the pull request:
Closes #20.

manoranjith commented on backend/ethereum/channel/adjudicator.go:
```diff
@@ -96,11 +96,12 @@ func (a *Adjudicator) callConclude(ctx context.Context, req channel.AdjudicatorR
 func (a *Adjudicator) callConcludeFinal(ctx context.Context, req channel.AdjudicatorReq) error {
-	return a.call(ctx, req, a.contract.ConcludeFinal)
+	err := a.call(ctx, req, a.contract.ConcludeFinal, "concludeFinal")
+	return err
```
Updated.

manoranjith commented on backend/ethereum/channel/contractbackend.go:
```diff
@@ -126,9 +131,14 @@ func (c *ContractBackend) NewTransactor(ctx context.Context, gasLimit uint64,
 func (c *ContractBackend) ConfirmTransaction(ctx context.Context, tx *types.Transaction, acc accounts.Account) (*types.Receipt, error) {
 	receipt, err := bind.WaitMined(ctx, c, tx)
 	if err != nil {
+		if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
```
Updated.

manoranjith commented on backend/ethereum/channel/deploy.go:
```diff
@@ -89,6 +101,9 @@ func deployContract(ctx context.Context, cb ContractBackend, deployer accounts.A
 		return common.Address{}, errors.WithMessage(err, "creating transaction")
 	}
 	if _, err := bind.WaitDeployed(ctx, cb, tx); err != nil {
+		if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
```
Updated.

manoranjith commented on backend/ethereum/channel/errors.go:
```diff
@@ -0,0 +1,45 @@
+// Copyright 2019 - See NOTICE file for copyright holders.
```
Updated.

manoranjith commented on backend/ethereum/channel/eth_depositor.go:
```diff
@@ -48,6 +50,7 @@ func (d *ETHDepositor) Deposit(ctx context.Context, req DepositReq) (types.Trans
 	opts.Value = req.Balance
 
 	tx, err := contract.Deposit(opts, req.FundingID, req.Balance)
+	fmt.Printf("\nSending deposit transaction for %s, fundingID: %x Balance: %s, TxID %s\n\n", req.Account.Address.Hex(), req.FundingID, req.Balance.String(), tx.Hash().Hex())
```
I had included it for debugging. Now it is removed.

manoranjith commented:
Hi @ggwpez 

It just occurred to me that, if we are adding the named error in the `backend/ethereum` package, then, just to check this error we will have to import the `backend/ethereum` pacakge. For that reason, do you think the definition of this error should be moved to the core ?

ggwpez commented:
Did not think about that before, thanks for the heads-up.  
Having this error in `channel` or `client` would be an API change since it would require all future backends to implement it correctly.  
The change would need to be documented in all public functions that can return it, sadly golang does not support us here.  
I think in this case it is easier to keep it in `backend/eth`, since the user could still unwrap the error and check from which backend it came.  
Maybe #21 would be a good candidate to think about extending our error API.

manoranjith commented:
> Did not think about that before, thanks for the heads-up.
> Having this error in `channel` or `client` would be an API change since it would require all future backends to implement it correctly.
> The change would need to be documented in all public functions that can return it, sadly golang does not support us here.
> I think in this case it is easier to keep it in `backend/eth`, since the user could still unwrap the error and check from which backend it came.
> Maybe #21 would be a good candidate to think about extending our error API.
> 
> PS: I do not see the changes from the last review, did you push it? @manoranjith

Yes, I agree this change need to be documented in all public functions. But I think in some cases, it would not be possible to import the corresponding backend package (in this case `ethereum`) just for checking the error. 

For example, in particular case of `perun-node`, we isolate the imports from `backend/ethereum` to a specific package that is used only during initialization of the state channel client in the `perun-node` program and use only the `core packages` of `go-perun` in all other parts. Because packages in `backend/ethereum` import code that is licensed under LGPL and hence we want to make a clear separation between code that uses dependencies licensed under LGPL and code that does not.

For the above reason, I think the usage of named error is also easier if we keep it in the core package.

manoranjith commented:
I have now force-pushed the changes. @ggwpez

ggwpez commented:
Ok that is a good argument. Then i think we could put `TxTimedoutError` into `channel/errors.go`.  
Do you have a strong opinion on that @matthiasgeihs ?

manoranjith commented:
I will then wait to hear from @matthiasgeihs on this.

matthiasgeihs commented:
Intuitively, I would have expected that these rather general error types fit well in the core packages. I am not sure about how much changes this requires, but I think @ggwpez already gave an indication.

Summary: I am fine with it, if there is no additional arguments against it.

manoranjith commented:
Okay @matthiasgeihs . In that case, I will update this PR with an additional commit to move the error to the core package. I will also update that this specific error will be returned in relevant scenario.

manoranjith commented:
I have moved the errors to [client] package and updated the comments for methods in [client] package that will return this error.

@ggwpez @matthiasgeihs Is there some other place where I should document the change in public API ?

manoranjith commented:
Rebased to latest dev.

matthiasgeihs commented on backend/ethereum/channel/adjudicator.go:
```diff
@@ -112,7 +113,7 @@ type adjFunc = func(
 // call calls the given contract function `fn` with the data from `req`.
 // `fn` should be a method of `a.contract`, like `a.contract.Register`.
-func (a *Adjudicator) call(ctx context.Context, req channel.AdjudicatorReq, fn adjFunc) error {
+func (a *Adjudicator) call(ctx context.Context, req channel.AdjudicatorReq, fn adjFunc, fnName string) error {
```
I think you should document the new parameter.

I don't love this manual way of providing the function name. I think it's not super reliable, because we may end up calling the same function under different names, if we are not careful. A more reliable way would be to introduce an enum type that ensures that you can only call certain functions with their name attached. Anyhow, it's not super important, but I wanted to point that out.

manoranjith replied:
Yeah, I agree that we should document this parameter.

Also, using enum for func names would be more reliable. Thanks for pointing it out. I will update the PR with the required change.

manoranjith replied:
I have updated the code to use enum definitions.

manoranjith replied:
In the [first update](https://github.com/hyperledger-labs/go-perun/compare/edda92c04127674699991b5c9ceb4d7ab8922103..d7cc92abeb076bc45ef9289acd13662ed54aefd5), I have added the enum in the core [client] package. But the it appeared to me like breaking the abstraction of the core package, because the specific on-chain transactions may depend upon the contract implementations for the blockchain backend.

So in the [following update](https://github.com/hyperledger-labs/go-perun/compare/d7cc92abeb076bc45ef9289acd13662ed54aefd5..04b1bd0bd5714299c1cff39186a6b5bf9e8855ee), I have moved the enum definitions to [backend/ethereum/channel] package. As i felt it belongs better here.

What are your thoughts on this ? @matthiasgeihs  @ggwpez

ggwpez reviewed (approved):
LGTM
